subprojects/soak/src/integTest/groovy/org/gradle/vfs/VirtualFileSystemRetentionSoakTest.groovy

```diff
@@ -96,17 +111,33 @@ class VirtualFileSystemRetentionSoakTest extends DaemonIntegrationSpec implement
         daemon.assertIdle()
 
         when:
-        NUMBER_OF_REPETITIONS.times { iteration ->
-            changeAllSourceFiles(iteration)
-            waitForChangesToBePickedUp()
+        numberOfChangeBatches.times { iteration ->
+            changeSourceFiles(iteration, numberOfChangedSourcesFilesPerBatch)
+            waitBetweenChangesToAvoidOverflow()
         }
+        boolean overflowDetected = detectOverflow(daemon)
+        int expectedNumberOfRetainedFiles = overflowDetected
+            ? 0
+            : retainedFilesInCurrentBuild - numberOfChangedSourcesFilesPerBatch
         then:
         succeeds("assemble")
         daemons.daemon.logFile == daemon.logFile
         daemon.assertIdle()
         assertWatchingSucceeded()
-        receivedFileSystemEvents >= minimumExpectedFileSystemEvents(TOTAL_NUMBER_OF_SOURCES, NUMBER_OF_REPETITIONS)
-        retainedFilesInCurrentBuild - TOTAL_NUMBER_OF_SOURCES == retainedFilesSinceLastBuild
+        receivedFileSystemEvents >= minimumExpectedFileSystemEvents(numberOfChangedSourcesFilesPerBatch, numberOfChangeBatches)
+        retainedFilesSinceLastBuild == expectedNumberOfRetainedFiles
+    }
+
+    private static boolean detectOverflow(DaemonFixture daemon) {
+        boolean overflowDetected = daemon.logContains(FileWatcherRegistry.Type.INVALIDATE.toString())
+        if (overflowDetected) {
+            println "Detected overflow in watcher, no files will be retained for the next build"
+        }
+        overflowDetected
+    }
+
+    private static void waitBetweenChangesToAvoidOverflow() {
+            Thread.sleep(150)
```

Comment:
I wonder if we could have some kine of debug printout that we could wait for here instead...

Reply:
What exactly should we wait here? Until the last file change is picked up by the Daemon? I don't think that is a good idea.

Reply:
Yeah, okay, let's not do it then.